### DISCUSSION TITLE
Issue with Displaying Post URLs after Overriding Post Retrieval

### DISCUSSION THREAD

**Discussion by 01Kuzma** [wintercms/wn-blog-plugin#56] [Q&A] [answered]
Hi there!
Having troubles with such functionality:
I have `intro.htm` which displays a hero carousel in three rows (made for a design purposes)
```
<div class="hero-section">
    {% partial 'hero-slider/hero1' posts=hero1Posts %}
    {% partial 'hero-slider/hero2' posts=hero2Posts %}
    {% partial 'hero-slider/hero3' posts=hero3Posts %}
</div>
```
Firstly, posts were displayed with the help of the `blogPosts` component in each hero1,2,3:
```
[viewBag]

[blogPosts]
pageNumber = "{{ :page }}"
postsPerPage = 14
noPostsMessage = "No posts found"
sortOrder = "random"
categoryPage = "blog-category"
postPage = "blog-post"
==
```

But there were many duplicates of posts in each row so I decided to display only unique posts in random order.
For this, the `intro.htm` has this PHP code:
```
use Winter\Blog\Models\Post;

function onStart() {
        
    $shuffledPosts = $this->shufflePosts();
    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];    
    
}

function shufflePosts() {
    $posts = Post::all();
    return $posts->shuffle();
}
```

Which pulls all posts, divides them into three parts and then I send each piece to the corresponding heroX view.
I removed the BlogPosts component from each hero view because it was overriding mine.

Each hero has the same code (I’ve removed unnecessary parts):
```
<!-- Hero section-1 -->
<div class="hero-slider owl-carousel">

    {% for post in posts %}    

    <div>                                      
            {% for image in post.featured_images|slice(0,1) %}
                {% if image %}
                    <!-- Local Featured Image -->
                    <div class="hero-item portfolio-item set-bg lazyload" data-setbg="{{ image.path | resize(600,600) }}">
                        <a href="{{ post.url }}" class="hero-link">
                            <h4>{{ post.title }}</h4>
                        </a>
                    </div>
                {% endif %}
            {% endfor %}
        {% endif %}
    </div>
    {% endfor %}
</div>
```

The problem: the new posts array doesn’t have a URL string, it’s null. The title, and images are displayed, but not not the `post.url`
The standard BlogPosts component did.
What I’m doing wrong and how to fix it?
Thanks!

**Comment by WebVPF** [comment, marked as answer]
In the Winter.Blog plugin, the Post model does not have a `url` field or property.

If you have a static entry path, you can use the `slug` field:

```twig
<a href="/path/{{ post.slug }}" class="hero-link">
    <h4>{{ post.title }}</h4>
</a>
```

---

The Post model also has a `getUrlParams` method, which returns an array of parameters for forming the URL https://github.com/wintercms/wn-blog-plugin/blob/4a14a17a607a6d7d90909e2679c99acfae76c47a/models/Post.php#L717-L740

You can use this data to create a URL for links in PHP code or do it like this in Twig:

```twig
{% set urlParams = post.getUrlParams() %}

<a href="/{{ urlParams.category }}/{{ urlParams.slug }}" class="hero-link">
    <h4>{{ post.title }}</h4>
</a>
```

I don’t use the Winter.Blog plugin, so need to check what data the method returns and how to form a path from it.

**Comment by 01Kuzma** [comment]
Thank you!
This does work!
Only the second provided solution throws me the error.
If I add `getUrlParams` method in `for` cycle:

```
    {% for post in posts  %}    
        {% set urlParams = post.getUrlParams() %}    
      ...
```
Throws this error:

```

Winter\Blog\Models\Post::getParamNameFromComponentProperty(): Argument #1 ($page) must be of type Cms\Classes\Page, null given, called in C:\laragon\www\winter3\plugins\winter\blog\models\Post.php on line 733

C:\laragon\www\winter3\plugins\winter\blog\traits\Urlable.php line 103


referencing to `protected function getParamNameFromComponentProperty(CmsPage $page, string $componentName, string $propertyName): ?string`
```

**Comment by WebVPF** [comment]
The error is due to the fact that when calling the method you need to specify a parameter - the CMS page that displays the post.

```twig
{% set urlParams = post.getUrlParams('here_is_the_post_cms_page') %}

<a href="/{{ urlParams.category }}/{{ urlParams.slug }}" class="hero-link">
    <h4>{{ post.title }}</h4>
</a>
```

**Comment by 01Kuzma** [comment]
Emm, still can't make it work :) Something is wrong on my side.
Those post links and images are displayed on an `intro.htm` page, while links lead to page `/post/:slug/?`
So, if I write `{% set urlParams = post.getUrlParams('post') %}   ` I get:
```
Winter\Blog\Models\Post::getUrlParams(): Argument #1 ($page) must be of type ?Cms\Classes\Page, string given, called in C:\laragon\www\winter3\vendor\twig\twig\src\Extension\CoreExtension.php on line 1637

from public function getUrlParams(?CmsPage $page = null): array
```

**Comment by LukeTowers** [comment]
The Posts component sets the url property here: https://github.com/wintercms/wn-blog-plugin/blob/main/components/Posts.php#L213-L222

Refer to that for how to properly set your URLs since you're interacting with the model instances directly instead of having the component take care of that for you.

**Comment by 01Kuzma** [comment]
@LukeTowers , you suggest to modify this code with properties from here ( [property ](https://github.com/wintercms/wn-blog-plugin/blob/main/components/Posts.php#L213-L222)) :
```
use Winter\Blog\Models\Post;

function onStart() {
        
    $shuffledPosts = $this->shufflePosts();
    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];       
    
}



function shufflePosts() {
    $posts = Post::all();   
   
    return $posts->shuffle();
}
```

and then call the `{% set urlParams = post.getUrlParams('post') %}` ?

**Reply by LukeTowers** [reply]
It should probably look something like this:

```php
use Winter\Blog\Models\Post;

function onStart() {
    $shuffledPosts = $this->shufflePosts();
    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];       
}

function shufflePosts() {
	// Note: you can use inRandomOrder() instead so that you can also limit the amount of data returned by the DB
    $posts = Post::all()->shuffle(); 
	foreach ($posts as $post)
		$post->setUrl('my/post-page', $this->controller);
	}
}
```

**Comment by 01Kuzma** [comment]
@LukeTowers , thank you!
I make it in this way:
```
use Winter\Blog\Models\Post;

function onStart() {
    $shuffledPosts = $this->shufflePosts();
    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];
    
    $this->setPostUrls($this['hero1Posts'], 'post');
    $this->setPostUrls($this['hero2Posts'], 'post');
    $this->setPostUrls($this['hero3Posts'], 'post');
}

function shufflePosts() {
    return Post::inRandomOrder()->get();
}

function setPostUrls($posts, $postPage) {
    $posts->each(function($post) use ($postPage) {
        $post->setUrl($postPage, $this->controller);  
    });
}

```
But in template the `{{ post.url }}` is empty

**Reply by LukeTowers** [reply]
Why aren't you just setting the URL on the shuffledPosts before chunking it?

**Reply by 01Kuzma** [reply]
@LukeTowers ,
Do you mean like this?

```php
function onStart() {
    $shuffledPosts = $this->shufflePosts();
    $this->setPostUrls($shuffledPosts, 'post');

    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];
}

function shufflePosts() {
    return Post::inRandomOrder()->get();
}

function setPostUrls($posts, $postPage) {
    $posts->each(function($post) use ($postPage) {
        $post->setUrl($postPage, $this->controller);  
    });
}
```

Still the same...

**Reply by 01Kuzma** [reply]
@LukeTowers , or it's a possible bug?

**Reply by LukeTowers** [reply]
Maybe try:

```php
function onStart() {
    $shuffledPosts = $this->shufflePosts();
    $chunkedPosts = $shuffledPosts->chunk(ceil(count($shuffledPosts) / 3));
    
    $this['hero1Posts'] = $chunkedPosts[0];    
    $this['hero2Posts'] = $chunkedPosts[1];
    $this['hero3Posts'] = $chunkedPosts[2];
}

function shufflePosts() {
    return $this->setPostUrls(Post::inRandomOrder()->get(), 'post');
}

function setPostUrls($posts, $postPage) {
    return $posts->each(function($post) use ($postPage) {
        $post->setUrl($postPage, $this->controller);  
    });
}

**Reply by 01Kuzma** [reply]
Still the same :(
URL is set to null
Here is a debug from `    $shuffledPosts = $this->shufflePosts();   dd($shuffledPosts);`:
![image](https://github.com/wintercms/wn-blog-plugin/assets/569504/cc92a5d8-b281-4bf4-bdd5-e12a31a00de1)

**Reply by LukeTowers** [reply]
Can you dd($this->controller)?

**Reply by 01Kuzma** [reply]
Sure, the
```
function setPostUrls($posts, $postPage) {    
    return $posts->each(function($post) use ($postPage) {
        dd($this->controller);
        $post->setUrl($postPage, $this->controller);  
    });
}
```
Gives:
![image](https://github.com/wintercms/wn-blog-plugin/assets/569504/64d61963-f022-4740-b1a5-924837e4d24f)

**Reply by LukeTowers** [reply]
Were you able to figure this out or are you still stuck?

**Reply by 01Kuzma** [reply]
No, still stuck :(

**Reply by 01Kuzma** [reply]
any news regarding this :) ?

**Reply by LukeTowers** [reply]
I can't remember @01Kuzma, did you send me your theme and posts data? I can try to replicate this sometime in the next week if you'd like.

**Comment by 01Kuzma** [comment]
> I can't remember @01Kuzma, did you send me your theme and posts data? I can try to replicate this sometime in the next week if you'd like.
No, I haven't sent you. Where could I send it to you? Do you need the file structure of the theme and the whole MySQL data?

**Reply by LukeTowers** [reply]
Yes, ideally. You can send it to hello@wintercms.com.

**Reply by 01Kuzma** [reply]
Done :)

**Reply by 01Kuzma** [reply]
Did you had time to look at this :) ?

**Reply by LukeTowers** [reply]
Not yet, sorry @01Kuzma. Were you able to figure it out yourself or do you still need me to take a look?

**Reply by 01Kuzma** [reply]
Hi,
I haven't  figure out.. Would be super nice if you look at it

**Comment by 01Kuzma** [comment]
Hi @LukeTowers ,
Did had you some time to investigate this issue?

**Reply by LukeTowers** [reply]
@01Kuzma this fell off my radar so I haven't looked at it. Have you looked into it any further?

**Reply by 01Kuzma** [reply]
> @01Kuzma this fell off my radar so I haven't looked at it. Have you looked into it any further?

I have tried several times, but was still stuck at the empty URL param :(
